File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} k
# @return {ListNode}
def reverse_k_group(head, k)
    temp=head
    result=[]
    while temp do
        next_next=temp.next
        temp.next=nil
        result.push(temp)
        temp=next_next
    end
    i=0
    while i+k<=result.length do
        reverse=result.slice(i,k).reverse
        l=0
        while l<reverse.length
            result[i+l]=reverse[l]
            l+=1
        end
        i+=k
    end
    top=result[0]
    temp=result[0]
    for i in 0..result.length-1
        temp.next=result[i+1]
        temp=temp.next
    end
    top
end

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    function reverse(start, end){
        if(start==end)
            return [end, end]
        let ret = reverse(start.next, end)
        ret[0].next=start
        return [start, end]
    }
    let result=null, left=head, right=0, temp=head, nex=null
    while(temp){
        right+=1
        if(right==k){
            let start=left, end=temp
            temp=temp.next
            let [tai, hea] = reverse(start, end)
                tai.next=null
            if(!result){
                result=hea
                nex=tai
            }else{
                nex.next=hea
                nex=tai
            }
            // console.log(right, result)
            left=temp
            right=0
        }else{
            temp=temp.next
        }
    }
    nex.next=left
    return result
};